### PR TITLE
Add mini profiles element

### PR DESCRIPTION
**Pull request by rhystmills**
## What does this change?
This PR adds rendering for the `MiniProfiles` element.

This is a variant of the [`List`](https://github.com/guardian/content-api-models/blob/ac9b9cb6826717a053667a80c5b06e543f8d669a/models/src/main/thrift/content/v1.thrift#L893) element from CAPI, in addition to [the Key Takeaways and Q&A elements](https://github.com/guardian/dotcom-rendering/pull/10795/files), which are rendered by DCR.

This PR involves:
- adding the new `MiniProfiles` variant to 'enhanceLists.ts'.
- representing the new element variant in DCR's models
- adding new components to represent the element
- adding new styles to 'palette.ts', mostly notably a bullet style for bullet lists in the `MiniProfiles` `bio` field, to be used in the `MiniProfile` component.

This work has been done according to some designs available in Figma:
https://www.figma.com/design/TK4X89VBAIZ65zmIqApOkI/Formats-%E2%80%94-Structured-elements?node-id=5-165675&t=BJK6VLoxp6kjbmuD-0

The MiniProfiles element has the following fields:

```ts
title: string;
body: Element[]
bio: string // HTML string representing rich text, including : `p`, `ul`, `li` block tags and `a`, `strong`, `em` inline tags.
endNote: string // Standard text, to be rendered as italic
```

## Why?
`MiniProfiles` are one of a number of new structure articles formats that Content Production have been developing, and which needs to be supported by our website.

## Screenshots

Non exhaustive variant screenshots, more available in stories:
### News
image

### Labs
image

### Sport
image

### Lifestyle
image

### CODE REVIEW

File: dotcom-rendering/src/components/MiniProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	margin: 0 0 2px 0;
+	border: none;
+	border-top: 4px solid ${palette('--heading-line')};
+`;
+
+// Nesting is necessary in the bio styles because we receive a string of html from the
+// field. This can contain the following tags:
+// Blocks: p, ul, li
+// Inline: strong, em, a
+const bioStyles = css`
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfile.tsx:31]: Using JSDoc will give better IDE integration:

```suggestion
/** Nesting is necessary in the bio styles because we receive a string of html from the
 * field. This can contain the following tags:
 * Blocks: `p`, `ul`, `li`
 * Inline: `strong`, `em`, `a`
 */
const bioStyles = css`
```

File: dotcom-rendering/src/components/MiniProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	${textSans14};
+	padding-top: 6px;
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfile.tsx:33]: Could we use [Source spaces](https://guardian.github.io/storybooks/?path=/docs/source_foundations-space--docs) here and throughout these styles?

File: dotcom-rendering/src/components/MiniProfile.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	color: ${palette('--mini-profiles-text-subdued')};
+	line-height: 1.3rem;
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfile.tsx:35]: Why is this necessary? Since the introduction of [specific text combinations like `textSans14`](https://github.com/guardian/csnx/pull/1335) we strongly discourage overriding line-height and other typography-related values.

File: dotcom-rendering/src/components/MiniProfiles.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+					pageId={pageId}
+					isAdFreeUser={isAdFreeUser}
+					isSensitive={isSensitive}
+					switches={switches}
+					abTests={abTests}
+					editionId={editionId}
+					hideCaption={hideCaption}
+					starRating={starRating}
+					key={index}
+					RenderArticleElement={RenderArticleElement}
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfiles.tsx:69]: The legacy React docs [recommend composition over inheritance](https://legacy.reactjs.org/docs/composition-vs-inheritance.html), and [Kent C. Dodds suggests so too](https://www.epicreact.dev/one-react-mistake-thats-slowing-you-down).

Would it be possible to update this to a single level like so:

```tsx
const MiniProfiles = (props: Props) => (
  <ol>
    {props.miniProfiles.map((miniProfile, index) => (
      <MiniProfileComponent>
        {miniProfile.body.map((element, elementIndex) => (
          <RenderArticleElement element={element} />
        ))}
      </MiniProfileComponent>
    ))}
  </ol>
);
```

File: dotcom-rendering/src/components/MiniProfile.tsx [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+					id={slugify(miniProfile.title)}
+					css={[subheadingStyles(format), headingMarginStyle]}
+				>
+					{miniProfile.title}
+				</h3>
+				<Bio html={miniProfile.bio} />
+				{miniProfile.body.map((element, index) => (
+					<RenderArticleElement
+						// eslint-disable-next-line react/no-array-index-key -- This is only rendered once so we can safely use index to suppress the warning
+						key={index}
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfile.tsx:137]: It’s worth noting that [`key={index}` is the React default](https://react.dev/learn/rendering-lists#rules-of-keys), so it’s not really adding anything to make it explicit.

Asserting that a component is only rendered once is also not a real guarantee – in strict mode while developing all elements are rendered twice, and if the mini profile’s body array were to change, it could trigger a re-render, so if there was a unique key it would be better.

File: dotcom-rendering/src/model/enhanceLists.ts [hunk 6]
```diff
@@ -32,6 +32,33 @@ const constructQAndAExplainer =
...
+		elements,
+		bio,
+		endNote,
+	}: {
+		title?: string;
+		elements: FEElement[];
+		bio?: string;
+		endNote?: string;
+	}) => {
+		// if the element is missing its title for any reason, we will skip it
```
Comment (mxdvl) [inline dotcom-rendering/src/model/enhanceLists.ts:48]: What about the empty string, i.e. `title.trim() === ""`?

File: dotcom-rendering/src/components/MiniProfile.tsx [hunk 7]
```diff
@@ -0,0 +1,186 @@
+import { css } from '@emotion/react';
+import { type ArticleFormat } from '@guardian/libs';
+import { neutral, textSans14 } from '@guardian/source/foundations';
+import type { EditionId } from '../lib/edition';
+import type { ArticleElementRenderer } from '../lib/renderElement';
+import { slugify } from '../model/enhance-H2s';
+import { sanitiseHTML } from '../model/sanitise';
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfile.tsx:7]: We actually [use `sanitize-html` thrice](https://github.com/search?q=repo%3Aguardian%2Fdotcom-rendering+%22from+%27sanitize-html%27%22&type=code) and [`sanitise.ts`’s sanitiseHTML only once](https://github.com/search?q=repo%3Aguardian%2Fdotcom-rendering+%22+sanitiseHTML+%22&type=code).

File: dotcom-rendering/src/components/MiniProfile.tsx [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+				css={bioStyles}
+				dangerouslySetInnerHTML={{ __html: sanitizedHtml }}
+			/>
+			<div css={bottomBorderStyles} />
+		</>
+	);
+};
+
+const EndNote = ({ text }: { text?: string }) => {
+	if (!text) return null;
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfile.tsx:180]: What about the empty string, i.e. `text.trim() === ""`?

**Review by mxdvl — APPROVED** [review]
Left a handful of comments that can be addressed in a follow-up.

File: dotcom-rendering/src/components/MiniProfile.tsx (hunk 5 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfile.tsx:137]: I didn't know this!

 In the case of the elements that are children of the MiniProfiles `body` - the order in the array should be determined by their order in CAPI so it should remain consistent on each render - it might be better to use that as an explanation. The elements don't have an `id` of any kind though, so without drilling down into what kind of element they are and doing a complicated case statement deriving a key for each one, I'm not sure how to derive one in this case. This will be the case for the other list elements as well

For the MiniProfiles themselves though, the `title` should normally be unique to each one.

File: dotcom-rendering/src/components/MiniProfiles.tsx (hunk 4 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfiles.tsx:69]: This is much nicer, cheers!

File: dotcom-rendering/src/components/MiniProfiles.tsx [hunk 9]
```diff
@@ -52,22 +52,35 @@ export const MiniProfiles = ({
 	return (
 		<ol data-ignore="global-ol-styling">
 			{miniProfiles.map((miniProfile, index) => (
+				// eslint-disable-next-line react/no-array-index-key -- Title should usually be identical, but in case it isn't, also use array index
```
Comment (mxdvl) [inline dotcom-rendering/src/components/MiniProfiles.tsx:55]: Is this necessary? If we pass a non-unique title, we’ll create an invalid HTML document as the H3’s id relies on its “slugification”.

Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfiles.tsx:55]: The title should be the name of the person the Profile is about, so editorially there shouldn't be a clash, unless two people in the 'Mini profile' article have the same name.

There's no 'hard' block in Composer or CAPI to having identical titles though.

File: dotcom-rendering/src/model/enhanceLists.ts (hunk 6 above)
Comment (rhystmills) [reply dotcom-rendering/src/model/enhanceLists.ts:48]: Good point, handled now.

File: dotcom-rendering/src/components/MiniProfile.tsx (hunk 1 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfile.tsx]: (Now done)

File: dotcom-rendering/src/components/MiniProfiles.tsx (hunk 9 above)
Comment (mxdvl) [reply dotcom-rendering/src/components/MiniProfiles.tsx:55]: Would [`generateId`](https://github.com/guardian/dotcom-rendering/blob/b15fbeff328cc7c1c8f969948b7b6fad5c0b8eed/dotcom-rendering/src/model/enhance-H2s.ts#L50) be the right tool for the job, then? It can be used to create a `key` and an `id`!

File: dotcom-rendering/src/components/MiniProfile.tsx (hunk 7 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfile.tsx:5]: I will follow the wisdom of the crowd.

File: dotcom-rendering/src/components/MiniProfile.tsx (hunk 8 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfile.tsx:133]: In this case I don't think we want to render anything if there is no text provided, whether that's `undefined` or `""`

File: dotcom-rendering/src/components/MiniProfiles.tsx (hunk 9 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfiles.tsx:55]: That looks like the right function for some of the elements, but I don't think will work for those that don't have an `html` field. I think we would need a function to handle them all.

What is missing for 'Mini profiles' will also be missing for Key Takeaways and Q&A though, so I'm happy to write one.

File: dotcom-rendering/src/components/MiniProfile.tsx (hunk 7 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfile.tsx]: (using `sanitise` from `sanitise-html` now)

File: dotcom-rendering/src/components/MiniProfile.tsx (hunk 3 above)
Comment (rhystmills) [reply dotcom-rendering/src/components/MiniProfile.tsx]: I'll avoid doing so, was working from a Figma file but happy to go in favour of consistency rather than design specifics that are probably flexible.
